tests/v2/validation/observability/stackstate_test.go

```diff
@@ -0,0 +1,338 @@
+	require.NoError(ss.T(), err)
+
+	ss.T().Log("Verifying the daemonsets of stackstate agent chart to have expected number of available replicas nodes")
+	err = extencharts.WatchAndWaitDaemonSets(client, ss.cluster.ID, charts.StackstateNamespace, meta.ListOptions{})
+	require.NoError(ss.T(), err)
+
+	stackstateAgentChartPreUpgrade, err := extencharts.GetChartStatus(client, ss.cluster.ID, charts.StackstateNamespace, charts.StackstateK8sAgent)
+
+	require.NoError(ss.T(), err)
+
+	// Validate current version of stackstate agent is one of the versions before latest
+	chartVersionPreUpgrade := stackstateAgentChartPreUpgrade.ChartDetails.Spec.Chart.Metadata.Version
+	require.Contains(ss.T(), versionsList[1:], chartVersionPreUpgrade)
+
+	ss.stackstateAgentInstallOptions.Version, err = client.Catalog.GetLatestChartVersion(charts.StackstateK8sAgent, rancherPartnerCharts)
+	require.NoError(ss.T(), err)
+
+	ss.T().Log("Upgrading stackstate agent chart to the latest version")
+	err = charts.UpgradeStackstateAgentChart(client, ss.stackstateAgentInstallOptions, ss.stackstateConfigs, systemProject.ID)
+	require.NoError(ss.T(), err)
+
+	ss.T().Log("Verifying the deployments of stackstate agent chart to have expected number of available replicas")
+	err = extencharts.WatchAndWaitDeployments(client, ss.cluster.ID, charts.StackstateNamespace, meta.ListOptions{})
+	require.NoError(ss.T(), err)
+
+	ss.T().Log("Verifying the daemonsets of stackstate agent chart to have expected number of available replicas nodes")
+	err = extencharts.WatchAndWaitDaemonSets(client, ss.cluster.ID, charts.StackstateNamespace, meta.ListOptions{})
+	require.NoError(ss.T(), err)
+
+	clusterObject, _, _ := extensionscluster.GetProvisioningClusterByName(ss.client, ss.client.RancherConfig.ClusterName, fleet.Namespace)
+	if clusterObject != nil {
+		status := &provv1.ClusterStatus{}
+		err := steveV1.ConvertToK8sType(clusterObject.Status, status)
+		require.NoError(ss.T(), err)
+
+		podErrors := pods.StatusPods(client, status.ClusterName)
+		require.Empty(ss.T(), podErrors)
+	}
+
+	stackstateAgentChartPostUpgrade, err := extencharts.GetChartStatus(client, ss.cluster.ID, charts.StackstateNamespace, charts.StackstateK8sAgent)
+	require.NoError(ss.T(), err)
+
+	ss.T().Log("Comparing installed and desired stackstate agent versions")
+	chartVersionPostUpgrade := stackstateAgentChartPostUpgrade.ChartDetails.Spec.Chart.Metadata.Version
+	require.Equal(ss.T(), ss.stackstateAgentInstallOptions.Version, chartVersionPostUpgrade)
+}
+
+func (ss *StackStateTestSuite) TestDynamicUpgradeStackstateAgentChart() {
+
+	subSession := ss.session.NewSession()
+	defer subSession.Cleanup()
+
+	client, err := ss.client.WithSession(subSession)
+	require.NoError(ss.T(), err)
+
+	versionToUpgrade := ss.stackstateConfigs.StackstateUpgradeVersion
+	if versionToUpgrade == "" {
+		ss.T().Skip("Skipping the test as no user version provided")
+	}
```

Comment:
why can't this just default to latest if stackstateConfigs.StackstateUpgradeVersion is nil?

Reply:
since this is a dynamic test and solely depends on user input I did not want to default it to latest as we already have an upgrade ss to latest static test.